Add register_module alias to nn.Module

rodrigoberriel opened the pull request:
Fixes #60397. I'm not sure how aliases are supposed to be implemented, but this is the most basic/direct way, IMO. As a side-effect, this implementation results in a "duplicate" doc entry, inheriting the one from `add_module`:

![monkey-patch](https://user-images.githubusercontent.com/7027770/133693137-8408d8e7-1f4f-436b-b176-57dda9bc3a32.png)

An alternative implementation could be:

```python
def register_module(self, name: str, module: Optional['Module']) -> None:
    r"""Alias for :func:`add_module`."""
    self.add_module(name, module)
```

which results in this documentation:

![image](https://user-images.githubusercontent.com/7027770/133693249-d969a71a-be44-489d-9633-4f38b44ab887.png)

Questions:
1. Should I replicate the tests? There are two for `add_module`: [test_add_module_raises_error_if_attr_exists](https://github.com/pytorch/pytorch/blob/873255c6d95342d144e9d1b633c16410844b934e/test/test_nn.py#L1420-L1434) and [test_add_module](https://github.com/pytorch/pytorch/blob/873255c6d95342d144e9d1b633c16410844b934e/test/test_nn.py#L1837-L1855).
2. This PR only adds `register_module` to `nn.Module`. There is an `add_module` in [`_RemoteModule`](https://github.com/pytorch/pytorch/blob/master/torch/distributed/nn/api/remote_module.py#L311-L312), which raises `NotSupported`, and there is another one in [`ConcreteModuleTypeBuilder`](https://github.com/pytorch/pytorch/blob/873255c6d95342d144e9d1b633c16410844b934e/torch/_C/__init__.pyi.in#L468), which means something else, I think. Should I do anything about them?

cc @ngimel @SsnL

rodrigoberriel committed:
Add register_module alias to nn.Module

rodrigoberriel committed:
Fix type hint

jbschlosser reviewed (commented):
> Fixes #60397. I'm not sure how aliases are supposed to be implemented, but this is the most basic/direct way, IMO. As a side-effect, this implementation results in a "duplicate" doc entry, inheriting the one from `add_module`:

I think this way looks good and it's nice that the docs are inherited.

> Questions:
> 
> 1. Should I replicate the tests? There are two for `add_module`: [test_add_module_raises_error_if_attr_exists](https://github.com/pytorch/pytorch/blob/873255c6d95342d144e9d1b633c16410844b934e/test/test_nn.py#L1420-L1434) and [test_add_module](https://github.com/pytorch/pytorch/blob/873255c6d95342d144e9d1b633c16410844b934e/test/test_nn.py#L1837-L1855).

Rather than duplicating the tests, maybe they could be slightly expanded to iterate over the two aliased functions with the the same test logic.

> 2. This PR only adds `register_module` to `nn.Module`. There is an `add_module` in [`_RemoteModule`](https://github.com/pytorch/pytorch/blob/master/torch/distributed/nn/api/remote_module.py#L311-L312), which raises `NotSupported`, and there is another one in [`ConcreteModuleTypeBuilder`](https://github.com/pytorch/pytorch/blob/873255c6d95342d144e9d1b633c16410844b934e/torch/_C/__init__.pyi.in#L468), which means something else, I think. Should I do anything about them?

cc @pritamdamania for `_RemoteModule`. I may be wrong, but I don't think you need to touch `ConcreteModuleTypeBuilder` since the `add_module` there is unrelated.

SsnL commented:
I think you should do the alternative impl, for 2 reasons:
1. the duplication of doc is confusing
2. things like `_RemoteModule` should be handled indeed, but you don't need to do that explicitly if you use the alternative impl.

rodrigoberriel committed:
Change to alternative implementation

rodrigoberriel committed:
Modify tests to iterate over the two aliased functions

rodrigoberriel commented:
I agree with @SsnL w.r.t. the implementation. I changed to the alternative, because although @jbschlosser showed support for the current one, he didn't say anything against the alternative :) Please, let me know if I should actually keep the current impl.

@jbschlosser I modified the tests to iterate over the aliases, but I'm not sure if this impl is what you were expecting.

As far as I understood, as `_RemoteModule` will inherit `register_module`, I don't need to explictly implement the alias there too. Therefore, I didn't change anything in `_RemoteModule`.

jbschlosser commented:
> I agree with @SsnL w.r.t. the implementation. I changed to the alternative, because although @jbschlosser showed support for the current one, he didn't say anything against the alternative :) Please, let me know if I should actually keep the current impl.

Cool, I'm good with the alternative as well.

> @jbschlosser I modified the tests to iterate over the aliases, but I'm not sure if this impl is what you were expecting.
> As far as I understood, as `_RemoteModule` will inherit `register_module`, I don't need to explictly implement the alias there too. Therefore, I didn't change anything in `_RemoteModule`.

Looks good!

jbschlosser reviewed (approved):
LGTM
